nibabel/streamlines/tractogram.py

```diff
@@ -0,0 +1,730 @@
+
+        Parameters
+        ----------
+        lazy : {False, True}, optional
+            If True, streamlines are *not* transformed in-place and a
+            :class:`LazyTractogram` object is returned. Otherwise, streamlines
+            are modified in-place.
+
+        Returns
+        -------
+        tractogram : :class:`Tractogram` or :class:`LazyTractogram` object
+            Tractogram where the streamlines have been sent to world space.
+            If the `lazy` option is true, it returns a :class:`LazyTractogram`
+            object, otherwise it returns a reference to this
+            :class:`Tractogram` object with updated streamlines.
+        """
+        if self.affine_to_rasmm is None:
+            msg = ("Streamlines are in a unknown space. This error can be"
+                   " avoided by setting the 'affine_to_rasmm' property.")
+            raise ValueError(msg)
+
+        return self.apply_affine(self.affine_to_rasmm, lazy=lazy)
+
+
+class LazyTractogram(Tractogram):
+    """ Lazy container for streamlines and their data information.
+
+    This container behaves lazily as it uses generator functions to manage
+    streamlines and their data information. This container is thus memory
+    friendly since it doesn't require having all those data loaded in memory.
+
+    Streamlines of a lazy tractogram can be in any coordinate system of your
+    choice as long as you provide the correct `affine_to_rasmm` matrix, at
+    construction time, that brings the streamlines back to *RAS+*, *mm* space,
+    where the coordinates (0,0,0) corresponds to the center of the voxel
+    (opposed to a corner).
+
+    Attributes
+    ----------
+    streamlines : generator function
+        Generator function yielding streamlines. Each streamline is an
+        ndarray of shape ($N_t$, 3) where $N_t$ is the number of points of
+        streamline $t$.
+    data_per_streamline : :class:`LazyDict` object
+        Dictionary where the items are (str, instantiated generator).
+        Each key represents an information $i$ to be kept along side every
+        streamline, and its associated value is a generator function
+        yielding that information via ndarrays of shape ($P_i$,) where
+        $P_i$ is the number scalar values to store for that particular
+        information $i$.
+    data_per_point : :class:`LazyDict` object
+        Dictionary where the items are (str, instantiated generator).
+        Each key represents an information $i$ to be kept along side every
+        point of every streamline, and its associated value is a generator
+        function yielding that information via ndarrays of shape
+        ($N_t$, $M_i$) where $N_t$ is the number of points for a particular
+        streamline $t$ and $M_i$ is the number scalar values to store for
+        that particular information $i$.
+
+    Notes
+    -----
+    LazyTractogram objects do not support indexing currently.
```

Comment:
So I guess they are mostly suited from linear operations, such as conversion or transformation, where you simply iterate once on each? If my understanding is correct, might be good to add a note on this in the Notes section.